### PR TITLE
Resolve only composite outputs early

### PR DESCRIPTION
**Pull request by lptr**
This is a followup to https://github.com/gradle/gradle/pull/22594.

The goal here is to make sure that we only call getters on `Task` objects early (=before they are needed to explore the task's build dependencies, or when snapshotting inputs and outputs) when necessary. After this change only two kinds of getter gets called early:

- `@Nested` getters,
- composite output getters, i.e. `@OutputDirectories` and `@OutputFiles`.

Both of these are necessary to determine the identity of a task's properties. All other getter calls happen lazily.

This is another step towards collecting a task's properties in a persistent data structure to be eventually finalized, instead of rediscovering them multiple times during the build.

### CODE REVIEW

**Comment by lptr** [comment]
@bot-gradle test QFL please.

File: subprojects/core/src/main/java/org/gradle/api/internal/tasks/properties/OutputFilePropertySpec.java [hunk 1]
```diff
@@ -18,9 +18,11 @@
 
 import org.gradle.internal.file.TreeType;
 
+import javax.annotation.Nullable;
 import java.io.File;
 
 public interface OutputFilePropertySpec extends FilePropertySpec {
     TreeType getOutputType();
+    @Nullable
```
Comment (wolfs) [inline subprojects/core/src/main/java/org/gradle/api/internal/tasks/properties/OutputFilePropertySpec.java:26]: Is the nullability which has been removed in https://github.com/gradle/gradle/pull/22594 now back?

File: subprojects/core/src/main/java/org/gradle/api/internal/tasks/properties/OutputUnpacker.java [hunk 2]
```diff
@@ -107,6 +109,23 @@ private static void resolveOutputFilePropertySpecs(
...
+        if (filePropertyType == OutputFilePropertyType.DIRECTORIES || filePropertyType == OutputFilePropertyType.FILES) {
+            // Eagerly unpack the value, because we need to create one property per root
+            Object unpackedValue = unpackValue(value, locationOnly);
+            if (unpackedValue == null) {
+                return;
+            }
+            resolveCompositeOutputFilePropertySpecs(ownerDisplayName, propertyName, unpackedValue, filePropertyType.getOutputType(), fileCollectionFactory, consumer);
+        } else {
+            FileCollectionInternal outputFiles = fileCollectionFactory.resolving((Callable<Object>) () -> unpackValue(value, locationOnly));
+            DefaultCacheableOutputFilePropertySpec filePropertySpec = new DefaultCacheableOutputFilePropertySpec(propertyName, null, outputFiles, filePropertyType.getOutputType());
```
Comment (wolfs) [inline subprojects/core/src/main/java/org/gradle/api/internal/tasks/properties/OutputUnpacker.java:121]: Is it now somewhere observable that the output files are resolved less often?

File: subprojects/core/src/main/java/org/gradle/api/internal/tasks/properties/OutputFilePropertySpec.java (hunk 1 above)
Comment (lptr) [reply subprojects/core/src/main/java/org/gradle/api/internal/tasks/properties/OutputFilePropertySpec.java:26]: Yep, I ended up reverting it indeed. However, it is now actually used again.

File: subprojects/core/src/main/java/org/gradle/api/internal/tasks/properties/OutputUnpacker.java (hunk 2 above)
Comment (lptr) [reply subprojects/core/src/main/java/org/gradle/api/internal/tasks/properties/OutputUnpacker.java:121]: No, we resolve them the same number of times for now, as evidenced by `TaskParametersIntegrationTest`.

File: subprojects/core/src/test/groovy/org/gradle/api/internal/project/taskfactory/AnnotationProcessingTaskFactoryTest.groovy [hunk 3]
```diff
@@ -796,9 +796,9 @@ class AnnotationProcessingTaskFactoryTest extends AbstractProjectBuilderSpec imp
         TaskWithOutputFiles                       | [[new File("some")]]           | []                                              | []                      | ["outputFiles\$1"]
         TaskWithOutputDir                         | [new File("some")]             | []                                              | []                      | ["outputDir"]
         TaskWithOutputDirs                        | [[new File("some")]]           | []                                              | []                      | ["outputDirs\$1"]
-        TaskWithOptionalOutputFile                | null                           | []                                              | []                      | []
+        TaskWithOptionalOutputFile                | null                           | []                                              | []                      | ["outputFile"]
```
Comment (wolfs) [inline subprojects/core/src/test/groovy/org/gradle/api/internal/project/taskfactory/AnnotationProcessingTaskFactoryTest.groovy:799]: Can this be a breaking change?

File: subprojects/core/src/main/java/org/gradle/api/internal/tasks/properties/OutputUnpacker.java [hunk 4]
```diff
@@ -107,6 +109,23 @@ private static void resolveOutputFilePropertySpecs(
...
     ) {
+        if (filePropertyType == OutputFilePropertyType.DIRECTORIES || filePropertyType == OutputFilePropertyType.FILES) {
+            // Eagerly unpack the value, because we need to create one property per root
+            Object unpackedValue = unpackValue(value, locationOnly);
+            if (unpackedValue == null) {
+                return;
+            }
+            resolveCompositeOutputFilePropertySpecs(ownerDisplayName, propertyName, unpackedValue, filePropertyType.getOutputType(), fileCollectionFactory, consumer);
+        } else {
+            FileCollectionInternal outputFiles = fileCollectionFactory.resolving((Callable<Object>) () -> unpackValue(value, locationOnly));
```
Comment (wolfs) [inline subprojects/core/src/main/java/org/gradle/api/internal/tasks/properties/OutputUnpacker.java:120]: I don't understand how that will help with finalization: `value` is the thing you want to finalize. Though this change puts it in a file collection as a callable which is not finalizable. How does this help?

File: subprojects/core/src/main/java/org/gradle/api/internal/tasks/properties/OutputUnpacker.java (hunk 2 above)
Comment (wolfs) [reply subprojects/core/src/main/java/org/gradle/api/internal/tasks/properties/OutputUnpacker.java:121]: Hm.... What does this late resolution mean then? Is there a way to exhibit the behavior in a test?

**Pull request closed without merging**